### PR TITLE
59743310 set the application id and pipeline_id

### PR DESCRIPTION
**Pull request by ke4**

### CODE REVIEW

File: lib/lims-api/context.rb [hunk 1]
```diff
@@ -11,6 +11,8 @@
 require 'lims-core/actions/bulk_action'
 require 'lims-core/helpers'
 
+require 'virtus'
+require 'aequitas/virtus_integration'
```
Comment (maxigit) [inline lib/lims-api/context.rb:15]: I don't think it's necessary, just stay consistent with the other attributes

File: lib/lims-api/context.rb [hunk 2]
```diff
@@ -27,27 +29,31 @@ module Api
...
         @store = store
         @last_session = nil
         @url_generator = url_generator
-        @application_id = application_id
+        @client_application_id = client_application_id
         @message_bus = message_bus
         super(content_type)
       end
 
+      attribute :user, String, :default => "", :required => false
```
Comment (maxigit) [inline lib/lims-api/context.rb:51]: stay consistent with the class, use attr_reader as the other attributes (or change everything to use attribute). On this occasion I prefer attr_reader

File: lib/lims-api/context.rb [hunk 3]
```diff
@@ -27,27 +29,31 @@ module Api
...
     class Context
+      include Virtus
+      include Aequitas
       include MimeTyped
 
       # Create a context with the specific store
       # @param [Lims::Core::Persistence::Store] store the store to retriev/store objects.
       # @param [Lims::Core::Persistence::MessageBus] bus to publish messages
-      # @param [String] application_id
+      # @param [String] client_application_id
```
Comment (maxigit) [inline lib/lims-api/context.rb:39]: Not sure we need to rename this. application_id fits with the actual Action#application

File: lib/lims-api/context.rb [hunk 4]
```diff
@@ -27,27 +29,31 @@ module Api
...
         @last_session = nil
         @url_generator = url_generator
-        @application_id = application_id
+        @client_application_id = client_application_id
         @message_bus = message_bus
         super(content_type)
       end
 
+      attribute :user, String, :default => "", :required => false
+      attribute :pipeline_id, String, :default => "", :required => false
```
Comment (maxigit) [inline lib/lims-api/context.rb:52]: shouldn't it be client_application_id ?

File: lib/lims-api/context.rb (hunk 2 above)
Comment (maxigit) [inline lib/lims-api/context.rb:51]: Why not set the default value there ?

File: lib/lims-api/context.rb [hunk 5]
```diff
@@ -184,6 +190,7 @@ def resource_class_for_class(klass)
           self.class.discover_resource_classes
           @@model_class_to_resource_class[klass]
         end
+        private :resource_class_for_class
```
Comment (maxigit) [inline lib/lims-api/context.rb:193]: Is it necessary ? (I mean is this modification parts of this story related to this pull request ?)

File: lib/lims-api/context.rb [hunk 6]
```diff
@@ -234,7 +241,10 @@ def execute_action(action)
         # @todo add user and 
         def create_action(action_class, attributes, resource_class=nil)
           resource_class ||= resource_class_for_class(action_class)
-          action = action_class.new( :store => store, :user => "user", :application => "application") do |a, session|
+          # TODO ke4 remove default values when client app sends these values
+          user = attributes["user"] || "user"
```
Comment (maxigit) [inline lib/lims-api/context.rb:245]: The point of having new member in the Context class is to not need user and pipeline_id in attributes. Just use the members.
pipeline_id should use @application_id.

File: lib/lims-api/context.rb (hunk 5 above)
Comment (ke4) [inline lib/lims-api/context.rb:193]: No it is not related to this story, but it was not in the right place and I didn't want to create a story for this small change.

On 30 Oct 2013, at 14:43, maxigit wrote:

> In lib/lims-api/context.rb:
> 
> > @@ -184,6 +190,7 @@ def resource_class_for_class(klass)
> >            self.class.discover_resource_classes
> >            @@model_class_to_resource_class[klass]
> >          end
> > -        private :resource_class_for_class
> >   Is it necessary ? (I mean is this modification parts of this story related to this pull request ?)
> 
> —
> Reply to this email directly or view it on GitHub.

## 

 The Wellcome Trust Sanger Institute is operated by Genome Research 
 Limited, a charity registered in England with number 1021457 and a 
 company registered in England with number 2742969, whose registered 
 office is 215 Euston Road, London, NW1 2BE.

File: lib/lims-api/context.rb [hunk 7]
```diff
@@ -344,7 +353,7 @@ def publish(action, resource)
             action = find_action_name(action) unless action.is_a? String
             payload = message_payload(action, resource)
             routing_key = resource.routing_key(action)
-            metadata = {:routing_key => routing_key, :app_id => @application_id}
+            metadata = {:routing_key => routing_key, :app_id => @message_bus.backend_application_id}
```
Comment (maxigit) [inline lib/lims-api/context.rb:356]: :app_id should be set my the MessageBus class itself.

File: lib/lims-api/context.rb [hunk 8]
```diff
@@ -355,7 +364,9 @@ def publish(action, resource)
           # @param [Hash] resource to publish
           def message_payload(action, resource)
             payload = Lims::Core::Helpers::load_json(resource.encoder_for(['application/json']).call)
-            payload.merge!({:action => action, :date => message_date, :user => "user"})
+            # TODO ke4 remove default value when client app sends these values
+            user ||= "user"
```
Comment (maxigit) [inline lib/lims-api/context.rb:368]: set the default value only once. (in initialize or at the attribute default value)

File: lib/lims-api/context_service.rb [hunk 9]
```diff
@@ -5,11 +5,11 @@ module Lims::Api
...
     # @param [Lims::Core::Persistence::Store] store  
     # @param [Lims::Core::Persistence::MessageBus] message bus
-    # @param [String,Nil] application_id
-    def initialize(store, message_bus, application_id=nil)
+    # @param [String,Nil] client_application_id
+    def initialize(store, message_bus, client_application_id=nil)
       @store = store
       @message_bus = message_bus
-      @application_id = application_id
+      @client_application_id = client_application_id
```
Comment (maxigit) [inline lib/lims-api/context_service.rb:12]: No needed. The client_application_id should be computed for each requests (ie each time we create a new Context)

File: lib/lims-api/resource.rb [hunk 10]
```diff
@@ -28,14 +28,14 @@ def actions
...
     # @param [String] action name
     # @return [String] routing key
-    def routing_key(for_action)
+    def routing_key(action)
       model = defined?(self.model_name) ? self.model_name : self.name
 
       MessageBus::generate_routing_key(
-        :pipeline_uuid => "pipeline",
-        :user_uuid => "user",
+        :pipeline_uuid => @context.pipeline_id,
```
Comment (maxigit) [inline lib/lims-api/resource.rb:35]: is it pipeline_id or application_id ???

**Pull request closed without merging**